internal/characters/tartaglia/abil.go

```diff
@@ -0,0 +1,411 @@
+	if !ok {
+		travel = 20
+	}
+
+	f, a := c.ActionFrames(core.ActionAttack, p)
+
+	if c.Core.Status.Duration("tartagliamelee") > 0 {
+		return c.meleeAttack(f, a)
+	}
+
+	d := c.Snapshot(
+		fmt.Sprintf("Normal %v", c.NormalCounter),
+		core.AttackTagNormal,
+		core.ICDTagNormalAttack,
+		core.ICDGroupDefault,
+		core.StrikeTypePierce,
+		core.Physical,
+		25,
+		attack[c.NormalCounter][c.TalentLvlAttack()],
+	)
+
+	c.QueueDmg(&d, f+travel)
+
+	c.AdvanceNormalIndex()
+
+	return f, a
+}
+
+// random delayed numbers
+var meleeDelayOffset = [][]int{
+	{0},
+	{0},
+	{0},
+	{0},
+	{0},
+	{1, 0},
+}
+
+// Melee stance attack.
+// Perform up to 6 consecutive Hydro strikes.
+func (c *char) meleeAttack(f, a int) (int, int) {
+	for i, mult := range eAttack[c.NormalCounter] {
+		c.AddTask(func() {
+			d := c.Snapshot(
+				fmt.Sprintf("Normal %v", c.NormalCounter),
+				core.AttackTagNormal,
+				core.ICDTagNormalAttack,
+				core.ICDGroupDefault,
+				core.StrikeTypeSlash,
+				core.Hydro,
+				25,
+				mult[c.TalentLvlSkill()],
+			)
+			d.OnHitCallback = c.rtSlashCallback
+			c.Core.Combat.ApplyDamage(&d)
+		}, "tartaglia-attack", f-meleeDelayOffset[c.NormalCounter][i])
+	}
+
+	c.AdvanceNormalIndex()
+	return f, a
+}
+
+//Once fully charged, deal Hydro DMG and apply the Riptide status.
+func (c *char) Aimed(p map[string]int) (int, int) {
+	f, a := c.ActionFrames(core.ActionAim, p)
+
+	travel, ok := p["travel"]
+	if !ok {
+		travel = 20
+	}
+	hitWeakPoint, ok := p["hitWeakPoint"]
+	if !ok {
+		hitWeakPoint = 0
+	}
+
+	d := c.Snapshot(
+		"Aim (Charged)",
+		core.AttackTagExtra,
+		core.ICDTagNone,
+		core.ICDGroupDefault,
+		core.StrikeTypePierce,
+		core.Hydro,
+		25,
+		aim[c.TalentLvlAttack()],
+	)
+
+	if hitWeakPoint != 0 {
+		d.HitWeakPoint = true
+	}
+	d.AnimationFrames = f
+	d.OnHitCallback = c.rtFlashCallback
+
+	c.QueueDmg(&d, travel+f)
+
+	return f, a
+}
+
+var meleeChargeDelayOffset = []int{
+	2, 0,
+}
+
+// since E is aoe, so this should be considered aoe too
+// hitWeakPoint: tartaglia can proc Prototype Cresent's Passive on Geovishap's weakspots.
+// Evidence: https://youtu.be/oOfeu5pW0oE
+func (c *char) ChargeAttack(p map[string]int) (int, int) {
+	f, a := c.ActionFrames(core.ActionCharge, p)
+
+	if c.Core.Status.Duration("tartagliamelee") == 0 {
+		return f, a
+	}
+
+	hitWeakPoint, ok := p["hitWeakPoint"]
+	if !ok {
+		hitWeakPoint = 0
+	}
+
+	// simply set 2 instances of CA to be snapshotted
+	for i, mult := range eCharge {
+		d := c.Snapshot(
+			"Charged Attack",
+			core.AttackTagExtra,
+			core.ICDTagExtraAttack,
+			core.ICDGroupDefault,
+			core.StrikeTypeSlash,
+			core.Hydro,
+			25,
+			mult[c.TalentLvlSkill()],
+		)
+		if hitWeakPoint != 0 {
+			d.HitWeakPoint = true
+		}
+		d.Targets = core.TargetAll
+		d.OnHitCallback = c.rtSlashCallback
+		c.QueueDmg(&d, f-meleeChargeDelayOffset[i])
+		// c.AddTask(func() {
+		// 	c.Core.Combat.ApplyDamage(&d)
+		// }, "tartaglia-charge-attack", f-meleeChargeDelayOffset[i])
+	}
```

Comment:
This is a dynamic ability, so you need to use either QueueDmgDynamic, or AddTask to align the snapshot creation and damage instance.